EIPS/eip-5289.md

```diff
@@ -0,0 +1,78 @@
+---
+eip: 5289
+title: Legal Smart Contracts
+description: Allows Smart Contracts to be Legally-Binding Off-Chain
+author: Pandapip1 (@Pandapip1)
+discussions-to: https://ethereum-magicians.org/t/discussions-to-pr-5289/9980
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-07-16
+requires: 165
+---
+
+## Abstract
+
+Currently, the real-world applications of smart contracts are limited by the fact that they aren't legally binding. This EIP proposes a standard that allows smart contracts to be legally binding by providing IPFS links to legal documents and ensuring that the users of the smart contract have privity with the relevant legal documents.
+
+(NOTE: None of the authors is a lawyer. If you are a lawyer and would like to contribute to this EIP, please contact Pandapip1 on Discord (Pandapip1#8943). This text will be removed by the time that this EIP goes into last call.)
+
+## Motivation
+
+NFTs have oftentimes been branded as a way to hold and prove copyright of a specific work. However, this, in practice, has almost never been the case. Most of the time, NFTs have no legally-binding meaning, and in the rare cases that do, the NFT simply provides a limited license for the initial holder to use the work (but cannot provide any license for any future holders).
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Legal Document Interface
+
+```solidity
+pragma solidity ^0.8.0;
+
+
+struct Multihash {
+    uint8 hashFunction;
+    uint8 size;
+    bytes32 digest;
+}
+
+interface ILegalDocument is IERC165 {
+    /// @notice The IPFS multihash of the legal document. This MUST resolve to a PDF.
+    function legalDocument() public pure returns (string memory);
+    
+    /// @notice The date the legal document was published (in days since January 1, 1970).
+    function datePublished() public pure returns (uint64);
+
+    /// @notice The ECDSA signature of the given user signing the IPFS hash.
+    function documentSignature(address user) public view returns (bytes memory);
```

Comment:
Since there is no way to iterate users, it means you need to get a user from somewhere else off-chain.  Why not just store the signature there instead of in contract storage?

Reply:
Why do users need to be iterated through?

Reply:
They don't, I'm just arguing that this function is unnecessary and it increases the cost of adding signatures to the system because it forces implementations to store the signatures on-chain unnecessarily.  Whatever mechanism an implementation has for finding a signer address, that same mechanism could be used to find the signer signature.

If this interface supported user iteration, then it would make more sense to have the signatures on-chain, because it would be reasonable to do signer discovery on-chain and (per above) the signatures should be where the user list is.

Reply:
As it happens, I replaced this method in a later version with one that only gives a timestamp. I agree with your assessment.